### DISCUSSION TITLE
@theia/git: missing filesystem notifications (create, delete,..)

### DISCUSSION THREAD

**Discussion by MikeKoda** [eclipse-theia/theia#14640] [Q&A] [answered]
lsp notifications didCreate, didDelete are sent when using the file explorer operations (create file, delete file). They are not sent when indirectly creating/deleting files by means of using the @theia/git ui and using the discard file changes operations (which in turn could create or delete files on the file system).

Any suggestions where to start looking at?

**Comment by msujew** [comment]
Hey @MikeKoda,

what version of Theia are you using? We recently noticed an issue in our file watching dependency that we fixed for 1.56.0, see https://github.com/eclipse-theia/theia/pull/14545. Since that's the system that's in charge of sending us any disk related events (something that we don't need if the user changes files via the explorer UI), that might be the cause of the issue. Updating to the latest version might resolve this issue for you as well.

**Comment by MikeKoda** [comment]
@msujew 

originally observed in 1.55. Switched to 1.56 and the new behaviour is as follows

OK: creating a new file via file explorer and deleting the file via git ui "discard changes" operation,  triggers the didDeleteFiles notification 
still NOK in 1.56: deleting a file from the local repo and restoring it via the git ui again does NOT trigger a didCreateFiles notification

**Reply by msujew** [reply]
cc @tsmaeder maybe that's interesting for you, since you're a bit deeper into it.

**Reply by tsmaeder** [reply]
@MikeKoda could you open an issue with steps to reproduce, please?

**Reply by MikeKoda** [reply]
@tsmaeder  https://github.com/eclipse-theia/theia/issues/14642

any suspect respectively where would you start looking at?

**Reply by tsmaeder** [reply]
@MikeKoda depends: how are you listening to the file changes? Where do you register? With the FileService?

**Reply by MikeKoda** [reply]
implicitly via this lsp client implementation 

~~https://github.com/microsoft/vscode-languageserver-node/blob/dadd73f7fc283b4d0adb602adadcf4be16ef3a7b/client/src/common/fileSystemWatcher.ts#L81~~

https://github.com/microsoft/vscode-languageserver-node/blob/dadd73f7fc283b4d0adb602adadcf4be16ef3a7b/client/src/common/fileOperations.ts#L299

**Comment by JonasHelming** [comment]
Please note that the @theia/git extension is actually deprecated (https://github.com/eclipse-theia/theia/issues/12745). However, I just noticed, it is not correctly marked as deprecated anymore, we will fix this.

**Reply by MikeKoda** [reply]
@JonasHelming 

which plugin do you suggest as a (drop-in) replacement and how does the migration path look like (i.e.  replacement for https://github.com/eclipse-theia/theia/tree/master/packages/git/src/common )? btw thanks for taking time to investigate and answering this

**Reply by JonasHelming** [reply]
The VS Code buil-in Git extension. It is part of our buil-in VS Code extension package anyways.

**Reply by MikeKoda** [reply]
@JonasHelming  sorry but how is it 'part of' or 'built-in' could you pls point me to some 'entry point' (src, docs, etc..) to start reading and learning about it by myself. THX for your answer and patience with me

add: a quick test of the combo  theia 1.57 with vscode.git 1.95.3 seems to introduce more bugs, when using the Discard Changes command from the git ui not only the lsp file create/delete notification are missing but also the file explorer view get stale in such a way that file delete/create are no visible automatically and it is required to use the 'Refresh Explorer' command manually

**Reply by JonasHelming** [reply]
No need to be sorry, this forum is for asking questions! :-)

Quick question: Do you use the Theia IDE as is or do you build your own tool based on Theia? I just want to understand your set-up to better answer the question.

**Reply by MikeKoda** [reply]
Lsp tooling with Runtime and Build Time dependency on  @theia/git

**Reply by JonasHelming** [reply]
I am not 100% sure what this means, but I assume you build a custom tooling based on Theia.

The built-in Git extension I was referring to is this: https://open-vsx.org/extension/vscode/git
Context: Theia is compatible with VS Code extensions. "Built-ins" is a pack of VS Code extensions that specific Theia-based products install by default, e.g. the Theia IDE includes these built-ins. The Git extension is one of these "built-in" VS Code extensions.

**Reply by MikeKoda** [reply]
@JonasHelming  @tsmaeder 

in the meantime..maybe it gets you an idea. 

-all events from the fileexplorer pane will be handled from [ExtHostFileSystemEventService.$onFileEvent](https://github.com/eclipse-theia/theia/blob/8d7fa9e9c56f2ac4a57abe3669e2032d1d40f5b9/packages/plugin-ext/src/plugin/file-system-event-service-ext-impl.ts#L170) and as well as [ExtHostFileSystemEventService .$onDidRunFileOperation](https://github.com/eclipse-theia/theia/blob/8d7fa9e9c56f2ac4a57abe3669e2032d1d40f5b9/packages/plugin-ext/src/plugin/file-system-event-service-ext-impl.ts#L176) i.e. they are handled twice?

-all file operation events triggered as a result from the git ui e.g. discard changes end up only here  [ExtHostFileSystemEventService.$onFileEvent](https://github.com/eclipse-theia/theia/blob/8d7fa9e9c56f2ac4a57abe3669e2032d1d40f5b9/packages/plugin-ext/src/plugin/file-system-event-service-ext-impl.ts#L170) 

-only the dispatched events from [ExtHostFileSystemEventService .$onDidRunFileOperation](https://github.com/eclipse-theia/theia/blob/8d7fa9e9c56f2ac4a57abe3669e2032d1d40f5b9/packages/plugin-ext/src/plugin/file-system-event-service-ext-impl.ts#L176) will end up in the respective handler code here e.g. [DidCreateFilesFeature](https://github.com/microsoft/vscode-languageserver-node/blob/dadd73f7fc283b4d0adb602adadcf4be16ef3a7b/client/src/common/fileOperations.ts#L298)

Are there some tests, to play around with, which could be used to verify the observation from above?

**Reply by msujew** [reply, marked as answer]
> only the dispatched events from [ExtHostFileSystemEventService .$onDidRunFileOperation](https://github.com/eclipse-theia/theia/blob/8d7fa9e9c56f2ac4a57abe3669e2032d1d40f5b9/packages/plugin-ext/src/plugin/file-system-event-service-ext-impl.ts#L176) will end up in the respective handler code here e.g. [DidCreateFilesFeature](https://github.com/microsoft/vscode-languageserver-node/blob/dadd73f7fc283b4d0adb602adadcf4be16ef3a7b/client/src/common/fileOperations.ts#L298)

Yes, but that's by design. I didn't quite notice that you explicitly meant those lsp events only. In the LSP, the editor is only supposed to send events that are explicitly triggered by the user in the IDE. For events that happen outside of the IDE, you need to use the [`workspace/didChangeWatchedFiles`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) notification. I believe those are reliably sent by Theia and VS Code as well.

Generally, if you notice that something doesn't work as expected in Theia with a VS Code extension, please confirm the behavior in VS Code first. We will only take action in case we deviate in behavior from VS Code.

**Reply by MikeKoda** [reply]
Hi @msujew,

first of all, thanks for also taking a look on that. Sry for my follow-up question(s) but did I understand you correctly that (workspace) file operations triggered from an extension like vscode.git considered outside of the IDE? If yes, could you pls describe me what is considered as inside and outside of the IDE means?

Regarding you request to confirm in VS Code first: Yes, I have time to do that as well but I would need some starting help from you on how to bootstrap such a setup. Could you pls point me to some code, tests, etc to see how I can setup and debug a browser based vscode including the vscode.git extension pls?

**Reply by msujew** [reply]
> Sry for my follow-up question(s) but did I understand you correctly that (workspace) file operations triggered from an extension like vscode.git considered outside of the IDE? If yes, could you pls describe me what is considered as inside and outside of the IDE means?

As far as I know, only file actions done through the explorer widget count as user-actions that will be forwarded to the language server via the `onDidCreate`/`onDidDelete` notifications. Everything else will be identified via our file watching mechanism which triggers the [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) notification.

> how I can setup and debug a browser based vscode including the vscode.git extension pls?

I don't think it's necessary to setup a browser based VS Code instance for that. You can just download and install [stock VS Code](https://code.visualstudio.com/) (which automatically includes `vscode.git`) and install the VS Code extension in there. Do you have access to the language server source code? Then you can open the extension source and debug the extension directly. That is usually the easiest way to debug any LSP issues.

**Reply by MikeKoda** [reply]
[didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) did it for my use case. Thanks for pointing this out.

**Reply by tsmaeder** [reply]
@MikeKoda could you close https://github.com/eclipse-theia/theia/issues/14642 if you think it's not an actual bug? Thx.
